### DISCUSSION TITLE
cannot pickle '_thread._local' object (type=type_error)

### DISCUSSION THREAD

**Discussion by detmoo** [pydantic/pydantic#4099] [Question] [answered]
Hello - 

Experiencing a problem on taking pydantic v1.9.1 over v1.9.0.

It is found that some previously working subclasses of `BaseModel` are now throwing `ValidationError` with message "cannot pickle '_thread._local' object (type=type_error)" 

In my case objects in question include attributes bound to a third party API and so cannot easily be recreated in a self-contained script (initial efforts with simplified scenario do not successfully recreate the issue). Nevertheless, below we have a debug excerpt of the problem as it is being experienced, where the offending object `AzDoRepo` is itself a child class of `BaseModel `. I would appreciate any best effort insight into which aspect of updates to 1.9.1 are possibly causing this to fall over (have exported the object that is failing to validate).

```
from pydantic import __version__, BaseModel, Field, ValidationError

from my.python.project import AzDoRepo  # this is a BaseModel object


class BazModel(BaseModel):
    repo: AzDoRepo = Field(..., description="problem")


def main():
    print('using pydantic version: ', __version__)
    repo = AzDoRepo.from_git_dir()  # this constructs from a context an instance of AzDoRepo model
    try:
        BazModel(repo=repo)
        print(f'Passed with pydantic=={__version__}')
    except ValidationError as e:
        print(f'Failed with pydantic=={__version__}: {e} \n')
        print('This is the object the causes trouble: \n', repo.dict(), '\n')


if __name__ == "__main__":
    main()
```

**Script Results**

_v1.9.1_

using pydantic version:  1.9.1
Failed with pydantic==1.9.1: 1 validation error for BazModel
repo
  cannot pickle '_thread._local' object (type=type_error)

This is the object the causes trouble:
 {'project_name': 'toolkits', 'org_name': 'myOrg', 'url': '/my.url', 'pat': 'xxxxx', 'repo_name': 'my.Repo.Name', '_connection': <azure.devops.connection.Connection object at 0x0000024E570D6AF0>, '_azdo_project_reference': None, '_azdo_project': None, '_AzDoBase__client': <azure.devops.released.git.git_client.GitClient object at 0x0000024E5AC7B1C0>, '_repo': None, '_build_repo': None}_

_v1.9.0_

using pydantic version:  1.9.0
Passed with pydantic==1.9.0

**Comment by samuelcolvin** [comment, marked as answer]
Duplicate of #4092

**Reply by detmoo** [reply]
Thank you for the attention. As per discussion on #4092/#4093 I can confirm that setting `Config.copy_on_model_validation = False` does result in successful validation of the model in question with v1.9.1. 

Apologies for the duplication! 

Since, on attempting to recreate my issue with a fully self-contained scenario I was unable to do so, I think it implies there are only certain characteristic(s) of a Model used as a Field that will cause the `copy_on_model_validation` to fall over.

**Reply by samuelcolvin** [reply]
Please use a hash on issue, PR and discussion IDs to make them links. This is very important to strengthen the knowledge web.

In summary, is this fixed by #4093?

**Reply by detmoo** [reply]
I think it is fair to say, in summary, this is fixed by #4093 as my scenario definitely concerns use of Models as Field types and furthermore the use of `Config.copy_on_model_validation = False` as discussed in #4092/#4093 is applicable/effective in restoring the desired behaviour. Accordingly, I would be happy for this discussion to be closed/resolved.

Nevertheless, under the summary, for now there remain two missing details from my understanding which may be of interest: (i) *to me* it is not clear the validation error "cannot pickle '_thread._local' object (type=type_error)" indicates a failure of `copy_on_model_validation` (ii) *to me* it is not clear what characteristics of a model would result in the failure of `copy_on_model_validation` when `Config.copy_on_model_validation = True', except to say that some Models will successfully copy when embedded as Fields, and others will produce a validation error on the parent Model.

Thanks once more. Unless you have further thoughts here I will just continue to track via #4093.
